test/state.coffee

```diff
@@ -0,0 +1,91 @@
+state = require('../lib/state')
+lineup = require('../lib/lineup')
+expect = require 'expect.js'
+
+# here we test manipulations of the lineup and window.location
+
+describe 'state', ->
+  actual = null
+  title = null
+  tests = [
+    # [pathname, locs, pages]
+    ['/', [], []],
+    ['/view/welcome-visitors', ['view'], ['welcome-visitors']],
+    ['/view/foo/view/bar', ['view', 'view'], ['foo', 'bar']],
+    ['/view/welcome-visitors/fed.wiki.org/featured-sites',
+      ['view', 'fed.wiki.org'], ['welcome-visitors', 'featured-sites']],
+    ['/welcome-visitors.html', ['view'], ['welcome-visitors']],
+    ['/how-to-wiki.html', ['view'], ['how-to-wiki']]
+  ]
+  before ->
+    global.$ = (el) -> {attr: (key) -> el[key]}
+    global.history =
+      pushState: (state, title, url) -> actual = url
+    lineup.bestTitle = () -> title
```

Comment:
We're stubbing jQuery's `$`, `history`, and the imported `lineup`. In specific test cases below we also stub `location`, and `document.title`. These allow the tests to pass where we do not have these browser APIs, but they will also be too tightly coupled to the current implementations.